### PR TITLE
Codechange: Use C++11 functions for threading

### PR DESCRIPTION
**Pull request by michicc**
The C++11 STL has functions for threading and synchronization, which means we can retire our homegrown platform-dependent code.

This PR assume a conforming C++11 compiler with conforming header files. DJGPP in it's current incarnation ships `thread` and `mutex` headers, but does not define the required types (not even as empty dummies). As such, this PR kills DOS unless a lot of #ifdef's are added.

We do not assume that starting a thread will succeeded, but do assume that if threads work, mutexes do work as well.

### CODE REVIEW

**Comment by TrueBrain** [comment]
I would suggest we remove DOS support for now, giving this PR priority. DOS has more issues (very slow), lacks networking, and more of that. Removing it has more benefits, but this shows how much we benefit from removing targets that are rarely used, and don't really support modern solutions. Just my 2 cents!

File: src/os/unix/unix.cpp [hunk 1]
```diff
@@ -43,11 +44,17 @@
...
 #if defined(__APPLE__)
-	#if defined(WITH_SDL)
+#	if defined(WITH_SDL)
 		/* the mac implementation needs this file included in the same file as main() */
-		#include <SDL.h>
-	#endif
+#		include <SDL.h>
+#	endif
+
+#	include "../os/macosx/macos.h"
```
Comment (glx22) [inline src/os/unix/unix.cpp]: already in os, should be ../macosx/macos.h

File: src/genworld.h [hunk 2]
```diff
@@ -61,7 +62,7 @@ struct GenWorldInfo {
 	uint size_y;           ///< Y-size of the map
 	GWDoneProc *proc;      ///< Proc that is called when done (can be NULL)
 	GWAbortProc *abortp;   ///< Proc that is called when aborting (can be NULL)
-	class ThreadObject *thread; ///< The thread we are in (can be NULL)
+	std::thread thread;    ///< The thread we are in (can be NULL)
```
Comment (LordAro) [inline src/genworld.h]: i imagine std::thread cannot be "NULL" ?

File: src/os/unix/unix.cpp [hunk 3]
```diff
@@ -317,4 +287,15 @@ void OSOpenBrowser(const char *url)
 	DEBUG(misc, 0, "Failed to open url: %s", url);
 	exit(0);
 }
-#endif
+#endif /* __APPLE__ */
+
+void SetCurrentThreadName(const char* threadName) {
```
Comment (LordAro) [inline src/os/unix/unix.cpp]:
```suggestion
void SetCurrentThreadName(const char *threadName)
{
```

File: src/os/windows/win32.cpp [hunk 4]
```diff
@@ -816,12 +803,12 @@ PACK_N(struct THREADNAME_INFO {
 /**
  * Signal thread name to any attached debuggers.
  */
-void SetWin32ThreadName(DWORD dwThreadID, const char* threadName)
+void SetCurrentThreadName(const char* threadName)
```
Comment (LordAro) [inline src/os/windows/win32.cpp]:
```suggestion
void SetCurrentThreadName(const char *threadName)
```

File: src/os/windows/win32.cpp [hunk 5]
```diff
@@ -832,4 +819,6 @@ void SetWin32ThreadName(DWORD dwThreadID, const char* threadName)
 	}
 #pragma warning(pop)
 }
+#else
+void SetCurrentThreadName(const char*) {}
```
Comment (LordAro) [inline src/os/windows/win32.cpp]:
```suggestion
void SetCurrentThreadName(const char *) {}
```
(I'm being really nitpicky now)

File: src/thread.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+ */
+void SetCurrentThreadName(const char *name);
+
+
+/**
+ * Start a new thread.
+ * @param thr Pointer to a thread object; may be \c NULL if a detached thread is wanted.
+ * @param name Name of the thread.
+ * @param _Fx Function to call on the thread.
+ * @param _Ax Arguments for the thread function.
```
Comment (LordAro) [inline src/thread.h:46]: @tparam

also, as per https://wiki.openttd.org/Coding_Style#Templates, template params should be `Tfoo`

File: src/thread.h [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+		if (thr != NULL) {
+			*thr = std::move(t);
+		} else {
+			t.detach();
+		}
+
+		return true;
+	} catch (const std::system_error&) {
+		/* Something went wrong, the system we are running on might not support threads. */
```
Comment (LordAro) [inline src/thread.h:73]: should say something here? at least a debug message, i'd expect

File: src/progress.h [hunk 8]
```diff
@@ -26,10 +26,20 @@ static inline bool HasModalProgress()
 	return _in_modal_progress;
 }
 
+/**
+ * Check if we can use a thread for modal progress.
+ * @return Threading usable?
+ */
+static inline bool UseThreadedModelProgress()
```
Comment (LordAro) [inline src/progress.h:33]: I wonder how useful this is, given not everywhere else checks for usable threads? (I only see this used for newgrf scan & worldgen) Should probably be all or nothing, and abort if the threads fail

**Review by LordAro — COMMENTED** [review]
Should resolve #6857 ?

File: src/progress.h (hunk 8 above)
Comment (michicc) [reply src/progress.h:33]: Constructing a mutex must succeed, locking a mutex may fail. The mutexes for model progress are locked right in `main`, if that fails it's instant exit. For the real existing C++11 implementations, mutex lock fail will only occur if threading at all is unusable.

The other threading sites try to create their threads first and fall back to a non-threaded implementation. We never got any bug reports that creating a thread succeeded whilst the subsequent mutex lock failed, and I don't see any reason that would change with this PR.

File: src/thread.h (hunk 7 above)
Comment (michicc) [reply src/thread.h:73]: We don't right now, but I can add one.

**Comment by michicc** [comment]
I don't see any reason why this PR should touch #6857. Internally, it all resolves down to pthreads anyway as before.

**Comment by orudge** [comment]
> I would suggest we remove DOS support for now, giving this PR priority. DOS has more issues (very slow), lacks networking, and more of that. Removing it has more benefits, but this shows how much we benefit from removing targets that are rarely used, and don't really support modern solutions. Just my 2 cents!

I did once get networking working on DOS via WATTCP, but that was a long time ago, and I never bothered trying to get that into OpenTTD itself. (In fact, I think it may have predated Rubidium's official DOS port.) No objections to dropping it, and you never know, perhaps one day somebody will 'port' std::thread to DJGPP!

**Review by TrueBrain — COMMENTED** [review]
Not really sure why you asked me for a review. I know nothing about C++11.

Minor update: we indeed dropped DOS support, so yeah .. that is solved ;)

File: src/saveload/saveload.cpp [hunk 9]
```diff
@@ -371,8 +372,8 @@ void NORETURN SlErrorCorruptFmt(const char *format, ...)
 
 
 typedef void (*AsyncSaveFinishProc)();                ///< Callback for when the savegame loading is finished.
-static AsyncSaveFinishProc _async_save_finish = NULL; ///< Callback to call when the savegame loading is finished.
-static ThreadObject *_save_thread;                    ///< The thread we're using to compress and write a savegame
+static std::atomic<AsyncSaveFinishProc> _async_save_finish; ///< Callback to call when the savegame loading is finished.
+static std::thread _save_thread;                      ///< The thread we're using to compress and write a savegame
```
Comment (PeterN) [inline src/saveload/saveload.cpp]: Can the comments be aligned?

**Comment by nielsmh** [comment]
This looked fine when I read through it a few days ago, are any more changes incoming or is it ready for review?

**Comment by michicc** [comment]
There were (and are) no changes except the comment aligning and rebasing.

File: src/music/extmidi.cpp [hunk 10]
```diff
@@ -151,7 +152,7 @@ void MusicDriver_ExtMidi::DoStop()
 			return;
 		}
 		/* Wait 10 milliseconds. */
-		CSleep(10);
+		std::this_thread::sleep_for(std::chrono::milliseconds(10));
```
Comment (nielsmh) [inline src/music/extmidi.cpp]: This replacement is so terribly long, couldn't the CSleep() function stay, maybe as an inline function in `thread.h` implemented via std::this_thread::sleep_for?

Comment (michicc) [reply src/music/extmidi.cpp]: It could. Matter of taste, I don't really care. Not sure what the others think.

Comment (michicc) [reply src/music/extmidi.cpp]: Done

**Review by PeterN — APPROVED** [review]

**Pull request merged**